Platform/OpenNtfsDxe/NTFS.c

```diff
@@ -0,0 +1,376 @@
+  // highest to lowest." Thus if NTFSVersion > FatVersion, it will somehow
+  // override Fat driver. FatVersion == 0xa.
+  //
+  .Version             = 0x09,
+  .ImageHandle         = NULL,
+  .DriverBindingHandle = NULL
+};
+
+EFI_STATUS
+EFIAPI
+NTFSEntryPoint (
+  IN EFI_HANDLE       ImageHandle,
+  IN EFI_SYSTEM_TABLE *SystemTable
+  )
+{
+  EFI_STATUS                Status;
+  EFI_LOADED_IMAGE_PROTOCOL *LoadedImage;
+
+  Status = gBS->OpenProtocol (
+      ImageHandle,
+      &gEfiLoadedImageProtocolGuid,
+      (VOID **) &LoadedImage,
+      ImageHandle,
+      NULL,
+      EFI_OPEN_PROTOCOL_GET_PROTOCOL
+      );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "NTFS: Could not open loaded image protocol\n"));
+    return Status;
+  }
+
+  gNTFSDriverBinding.ImageHandle = ImageHandle;
+  gNTFSDriverBinding.DriverBindingHandle = ImageHandle;
+
+  Status = gBS->InstallMultipleProtocolInterfaces (
+      &ImageHandle,
+      &gEfiDriverBindingProtocolGuid,
+      &gNTFSDriverBinding,
+      &gEfiComponentNameProtocolGuid,
+      &gNTFSDriverNames,
+      NULL
+      );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "NTFS: Could not bind driver\n"));
+    return Status;
+  }
+
+  LoadedImage->Unload = UnloadNTFSDriver;
+
+  return Status;
+}
+
+EFI_STATUS
+EFIAPI
+UnloadNTFSDriver (
+  IN EFI_HANDLE ImageHandle
+  )
+{
+  EFI_HANDLE *Buffer;
+  UINTN      NoOfHandles;
+  UINTN      Index;
+  EFI_STATUS Status;
+
+  Status = gBS->LocateHandleBuffer (
+    AllHandles,
+    NULL,
+    NULL,
+    &NoOfHandles,
+    &Buffer
+    );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  for (Index = 0; Index < NoOfHandles; Index++) {
+    gBS->DisconnectController (Buffer[Index], ImageHandle, NULL);
+  }
+
+  if (Buffer != NULL) {
+    FreePool(Buffer);
+  }
+
+  Status = gBS->UninstallMultipleProtocolInterfaces (
+    ImageHandle,
+    &gEfiDriverBindingProtocolGuid,
+    &gNTFSDriverBinding,
+    &gEfiComponentNameProtocolGuid,
+    &gNTFSDriverNames,
+    NULL
+    );
+
+  return Status;
+}
+
+EFI_STATUS
+EFIAPI
+NTFSSupported (
+  IN EFI_DRIVER_BINDING_PROTOCOL *This,
+  IN EFI_HANDLE                  Controller,
+  IN EFI_DEVICE_PATH_PROTOCOL    *RemainingDevicePath
+  )
+{
+  EFI_STATUS            Status;
+  EFI_DISK_IO_PROTOCOL  *DiskIo;
+
+  Status = gBS->OpenProtocol (
+    Controller,
+    &gEfiDiskIoProtocolGuid,
+    (VOID **)&DiskIo,
+    This->DriverBindingHandle,
+    Controller,
+    EFI_OPEN_PROTOCOL_BY_DRIVER
+    );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = gBS->CloseProtocol (
+    Controller,
+    &gEfiDiskIoProtocolGuid,
+    This->DriverBindingHandle,
+    Controller
+    );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+EFIAPI
+OpenVolume (
+  IN EFI_SIMPLE_FILE_SYSTEM_PROTOCOL *This,
+  OUT EFI_FILE_PROTOCOL              **Root
+  )
+{
+  EFI_FS         *Instance;
+
+  ASSERT (This != NULL);
+
+  Instance = (EFI_FS *) This;
```

Comment:
Such casts is not easy to verify at first sight (and may break when the struct layout is changed), any such should use ``CR``